File: src/Tiedragon.XmppMessenger.Core/Xmpp/XmppPubSubAnnouncements.cs
using System.Globalization;
using System.Xml.Linq;

namespace Tiedragon.XmppMessenger.Core.Xmpp;

public static class XmppPubSubAnnouncements
{
    public const string DefaultNode = "urn:tiedragon:teletyptel:announcements";

    public const string AtomNamespaceName = "http://www.w3.org/2005/Atom";

    public const string PriorityScheme = "urn:tiedragon:teletyptel:announcement-priority";

    public static XmppIq CreatePublishRequest(
        string id,
        XmppAnnouncement announcement,
        string node = DefaultNode,
        XmppAddress? service = null)
    {
        ArgumentNullException.ThrowIfNull(announcement);
        return XmppPersonalEventing.CreatePublishRequest(
            id,
            node,
            announcement.Id,
            CreateAtomEntry(announcement),
            service);
    }

    public static XmppIq CreateItemsRequest(
        string id,
        XmppAddress? service,
        string node = DefaultNode,
        int? maxItems = null)
    {
        return XmppPersonalEventing.CreateItemsRequest(id, node, service, maxItems: maxItems);
    }

    public static XmppIq CreateSubscribeRequest(
        string id,
        XmppAddress jid,
        XmppAddress service,
        string node = DefaultNode)
    {
        return XmppPubSub.CreateSubscribeRequest(id, node, jid, service);
    }

    public static bool TryParseItems(
        XmppIq iq,
        out IReadOnlyList<XmppAnnouncement> announcements,
        string node = DefaultNode)
    {
        announcements = [];
        if (!XmppPersonalEventing.TryParseItemsResult(iq, out var items)
            || items is null
            || !string.Equals(items.Node, node, StringComparison.Ordinal))
        {
            return false;
        }

        announcements = items.Items
            .SelectMany(item => item.Payloads)
            .Select(payload => TryParseAtomEntry(payload, out var announcement) ? announcement : null)
            .Where(announcement => announcement is not null)
            .Cast<XmppAnnouncement>()
            .ToArray();
        return true;
    }

    public static bool TryParseNotification(
        XElement message,
        out IReadOnlyList<XmppAnnouncement> announcements,
        string node = DefaultNode)
    {
        announcements = [];
        if (!XmppPersonalEventing.TryParseNotification(message, out var notification)
            || notification is null)
        {
            return false;
        }

        announcements = notification
            .ForNode(node)
            .SelectMany(node => node.Items)
            .SelectMany(item => item.Payloads)
            .Select(payload => TryParseAtomEntry(payload, out var announcement) ? announcement : null)
            .Where(announcement => announcement is not null)
            .Cast<XmppAnnouncement>()
            .ToArray();
        return announcements.Count > 0;
    }

    public static XElement CreateAtomEntry(XmppAnnouncement announcement)
    {
        ArgumentNullException.ThrowIfNull(announcement);
        ArgumentException.ThrowIfNullOrWhiteSpace(announcement.Id);
        ArgumentException.ThrowIfNullOrWhiteSpace(announcement.Title);
        ArgumentException.ThrowIfNullOrWhiteSpace(announcement.Summary);

        var atom = XNamespace.Get(AtomNamespaceName);
        var entry = new XElement(
            atom + "entry",
            new XElement(atom + "id", announcement.Id),
            new XElement(atom + "title", announcement.Title),
            new XElement(atom + "summary", announcement.Summary));

        if (!string.IsNullOrWhiteSpace(announcement.Language))
        {
            entry.SetAttributeValue(XNamespace.Xml + "lang", announcement.Language);
        }

        if (announcement.Published is not null)
        {
            entry.Add(new XElement(atom + "published", FormatDate(announcement.Published.Value)));
        }

        entry.Add(new XElement(atom + "updated", FormatDate(announcement.Updated ?? announcement.Published ?? DateTimeOffset.UtcNow)));

        if (announcement.Link is not null)
        {
            entry.Add(new XElement(atom + "link", new XAttribute("href", announcement.Link.AbsoluteUri)));
        }

        if (!string.IsNullOrWhiteSpace(announcement.Category))
        {
            entry.Add(new XElement(atom + "category", new XAttribute("term", announcement.Category)));
        }

        if (!string.IsNullOrWhiteSpace(announcement.Priority))
        {
            entry.Add(new XElement(
                atom + "category",
                new XAttribute("scheme", PriorityScheme),
                new XAttribute("term", announcement.Priority)));
        }

        return entry;
    }

    public static bool TryParseAtomEntry(XElement element, out XmppAnnouncement? announcement)
    {
        announcement = null;
        var atom = XNamespace.Get(AtomNamespaceName);
        if (element.Name != atom + "entry")
        {
            return false;
        }

        var id = element.Element(atom + "id")?.Value;
        var title = element.Element(atom + "title")?.Value;
        var summary = element.Element(atom + "summary")?.Value;
        if (string.IsNullOrWhiteSpace(id)
            || string.IsNullOrWhiteSpace(title)
            || string.IsNullOrWhiteSpace(summary))
        {
            return false;
        }

        Uri.TryCreate((string?)element.Element(atom + "link")?.Attribute("href"), UriKind.Absolute, out var link);
        DateTimeOffset.TryParse(
            element.Element(atom + "published")?.Value,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal,
            out var published);
        DateTimeOffset.TryParse(
            element.Element(atom + "updated")?.Value,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal,
            out var updated);

        announcement = new XmppAnnouncement(
            Id: id,
            Title: title,
            Summary: summary,
            Link: link,
            Language: (string?)element.Attribute(XNamespace.Xml + "lang"),
            Category: element.Elements(atom + "category")
                .FirstOrDefault(category => category.Attribute("scheme") is null)
                ?.Attribute("term")
                ?.Value,
            Priority: element.Elements(atom + "category")
                .FirstOrDefault(category => string.Equals(
                    (string?)category.Attribute("scheme"),
                    PriorityScheme,
                    StringComparison.Ordinal))
                ?.Attribute("term")
                ?.Value,
            Published: published == default ? null : published,
            Updated: updated == default ? null : updated);
        return true;
    }

    private static string FormatDate(DateTimeOffset value)
    {
        return value.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture);
    }
}

public sealed record XmppAnnouncement(
    string Id,
    string Title,
    string Summary,
    Uri? Link = null,
    string? Language = null,
    string? Category = null,
    string? Priority = null,
    DateTimeOffset? Published = null,
    DateTimeOffset? Updated = null);

File: php/lib/Xmpp/XmppPubSubAnnouncements.php
<?php
declare(strict_types=1);

namespace Tiedragon\Xmpp;

use DOMElement;

final class XmppPubSubAnnouncements
{
    public const DEFAULT_NODE = 'urn:tiedragon:teletyptel:announcements';
    public const ATOM_NS = 'http://www.w3.org/2005/Atom';
    public const PRIORITY_SCHEME = 'urn:tiedragon:teletyptel:announcement-priority';

    /**
     * @param array{
     *   id:string,
     *   title:string,
     *   summary:string,
     *   link?:string|null,
     *   language?:string|null,
     *   category?:string|null,
     *   priority?:string|null,
     *   published?:string|\DateTimeInterface|null,
     *   updated?:string|\DateTimeInterface|null
     * } $announcement
     */
    public static function publishRequest(
        string $id,
        array $announcement,
        string $node = self::DEFAULT_NODE,
        XmppJid|string|null $service = null
    ): string {
        return XmppPersonalEventing::publishRequest($id, $node, $announcement['id'], self::atomEntry($announcement), $service);
    }

    public static function itemsRequest(
        string $id,
        XmppJid|string|null $service = null,
        string $node = self::DEFAULT_NODE,
        ?int $maxItems = null
    ): string {
        return XmppPersonalEventing::itemsRequest($id, $node, $service, maxItems: $maxItems);
    }

    public static function subscribeRequest(
        string $id,
        XmppJid|string $jid,
        XmppJid|string $service,
        string $node = self::DEFAULT_NODE
    ): string {
        return XmppPubSub::subscribeRequest($id, $node, $jid, $service);
    }

    /**
     * @param array<string,mixed> $announcement
     */
    public static function atomEntry(array $announcement): string
    {
        foreach (['id', 'title', 'summary'] as $required) {
            if (trim((string)($announcement[$required] ?? '')) === '') {
                throw new \InvalidArgumentException("Announcement {$required} is required.");
            }
        }

        $published = self::dateValue($announcement['published'] ?? null);
        $updated = self::dateValue($announcement['updated'] ?? null) ?? $published ?? gmdate('c');
        $xml = '<entry xmlns="' . self::ATOM_NS . '"'
            . XmppXml::attributes(['xml:lang' => $announcement['language'] ?? null])
            . '>'
            . XmppXml::textElement('id', self::ATOM_NS, (string)$announcement['id'])
            . XmppXml::textElement('title', self::ATOM_NS, (string)$announcement['title'])
            . XmppXml::textElement('summary', self::ATOM_NS, (string)$announcement['summary']);

        if ($published !== null) {
            $xml .= XmppXml::textElement('published', self::ATOM_NS, $published);
        }
        $xml .= XmppXml::textElement('updated', self::ATOM_NS, $updated);

        if (!empty($announcement['link'])) {
            $xml .= '<link' . XmppXml::attributes(['xmlns' => self::ATOM_NS, 'href' => (string)$announcement['link']]) . '/>';
        }
        if (!empty($announcement['category'])) {
            $xml .= '<category' . XmppXml::attributes(['xmlns' => self::ATOM_NS, 'term' => (string)$announcement['category']]) . '/>';
        }
        if (!empty($announcement['priority'])) {
            $xml .= '<category' . XmppXml::attributes([
                'xmlns' => self::ATOM_NS,
                'scheme' => self::PRIORITY_SCHEME,
                'term' => (string)$announcement['priority'],
            ]) . '/>';
        }

        return $xml . '</entry>';
    }

    /**
     * @return list<array{id:string,title:string,summary:string,link:?string,language:?string,category:?string,priority:?string,published:?string,updated:?string}>
     */
    public static function parseItems(string $xml, string $node = self::DEFAULT_NODE): array
    {
        $items = XmppPersonalEventing::parseItemsResult($xml);
        if ($items === null || $items['node'] !== $node) {
            return [];
        }

        return self::parsePayloads($items['items']);
    }

    /**
     * @return list<array{id:string,title:string,summary:string,link:?string,language:?string,category:?string,priority:?string,published:?string,updated:?string}>
     */
    public static function parseNotification(string $xml, string $node = self::DEFAULT_NODE): array
    {
        $event = XmppPersonalEventing::parseNotification($xml);
        if ($event === null || $event['node'] !== $node) {
            return [];
        }

        return self::parsePayloads($event['items']);
    }

    /**
     * @return array{id:string,title:string,summary:string,link:?string,language:?string,category:?string,priority:?string,published:?string,updated:?string}|null
     */
    public static function parseAtomEntry(string $xml): ?array
    {
        $document = XmppXml::document($xml);
        $root = $document->documentElement;
        if (!$root instanceof DOMElement || $root->namespaceURI !== self::ATOM_NS || $root->localName !== 'entry') {
            return null;
        }

        $id = XmppXml::firstElementText($root, self::ATOM_NS, 'id');
        $title = XmppXml::firstElementText($root, self::ATOM_NS, 'title');
        $summary = XmppXml::firstElementText($root, self::ATOM_NS, 'summary');
        if ($id === null || $title === null || $summary === null) {
            return null;
        }

        $link = null;
        $category = null;
        $priority = null;
        foreach ($root->childNodes as $child) {
            if (!$child instanceof DOMElement || $child->namespaceURI !== self::ATOM_NS) {
                continue;
            }
            if ($child->localName === 'link' && $link === null) {
                $link = $child->getAttribute('href') ?: null;
            } elseif ($child->localName === 'category') {
                if ($child->getAttribute('scheme') === self::PRIORITY_SCHEME) {
                    $priority = $child->getAttribute('term') ?: null;
                } elseif ($category === null) {
                    $category = $child->getAttribute('term') ?: null;
                }
            }
        }

        return [
            'id' => $id,
            'title' => $title,
            'summary' => $summary,
            'link' => $link,
            'language' => $root->getAttribute('xml:lang') ?: null,
            'category' => $category,
            'priority' => $priority,
            'published' => XmppXml::firstElementText($root, self::ATOM_NS, 'published'),
            'updated' => XmppXml::firstElementText($root, self::ATOM_NS, 'updated'),
        ];
    }

    /**
     * @param list<array{id:?string,publisher:?string,payload:string}> $items
     * @return list<array{id:string,title:string,summary:string,link:?string,language:?string,category:?string,priority:?string,published:?string,updated:?string}>
     */
    private static function parsePayloads(array $items): array
    {
        $announcements = [];
        foreach ($items as $item) {
            $announcement = self::parseAtomEntry($item['payload']);
            if ($announcement !== null) {
                $announcements[] = $announcement;
            }
        }

        return $announcements;
    }

    private static function dateValue(mixed $value): ?string
    {
        if ($value instanceof \DateTimeInterface) {
            return $value->format(\DateTimeInterface::ATOM);
        }

        return $value === null || $value === '' ? null : (string)$value;
    }
}
